pkg/vsphere/tasks/waiter.go

```diff
@@ -89,6 +102,56 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
 	}
 }
 
+func fixTask(ctx context.Context, sess *session.Session, target object.Reference) error {
+	switch vmm := target.(type) {
+	case *object.VirtualMachine:
+		vm := vm.NewVirtualMachine(ctx, sess, target.Reference())
+		err := vm.FixInvalidState(ctx)
+		// make sure original object is updated, to make sure task retry works
+		vmm.Common = vm.Common
+		return err
+	case *vm.VirtualMachine:
+		return vmm.FixInvalidState(ctx)
+	default:
+		return errors.New("Unable to fix non-vm object")
+	}
+}
+
+// check if task is in progress, or vm is in invalid state.
+func needsRetry(target object.Reference, err error) bool {
+	if isTaskInProgress(err) {
+		return true
+	}
+	if needsFix(target, err) {
+		return true
+	}
+	return false
+}
+
+func needsFix(target object.Reference, err error) bool {
+	f, ok := err.(types.HasFault)
+	if !ok {
+		return false
+	}
+	switch f.Fault().(type) {
+	case *types.InvalidState:
+	default:
+		log.Debugf("Do not fix non invalid state error")
+		return false
+	}
+	if target == nil {
```

Comment:
Initially I'm doing that way, but I think invalid state will be a much more rare condition than any others. In fact, we'll never have a nil target, unless in test case.